spark/src/main/java/com/netflix/iceberg/spark/data/SparkParquetWriters.java

```diff
@@ -0,0 +1,447 @@
+          "Cannot write value as decimal(%s,%s), wrong scale: %s", precision, scale, decimal);
+      Preconditions.checkArgument(decimal.precision() <= precision,
+          "Cannot write value as decimal(%s,%s), too large: %s", precision, scale, decimal);
+
+      BigDecimal bigDecimal = decimal.toJavaBigDecimal();
+
+      byte fillByte = (byte) (bigDecimal.signum() < 0 ? 0xFF : 0x00);
+      byte[] unscaled = bigDecimal.unscaledValue().toByteArray();
+      byte[] buf = bytes.get();
+      int offset = length - unscaled.length;
+
+      for (int i = 0; i < length; i += 1) {
+        if (i < offset) {
+          buf[i] = fillByte;
+        } else {
+          buf[i] = unscaled[i - offset];
+        }
+      }
+
+      column.writeBinary(repetitionLevel, Binary.fromReusedByteArray(buf));
+    }
+  }
+
+  private static class ByteArrayWriter extends PrimitiveWriter<byte[]> {
+    private ByteArrayWriter(ColumnDescriptor desc) {
+      super(desc);
+    }
+
+    @Override
+    public void write(int repetitionLevel, byte[] bytes) {
+      column.writeBinary(repetitionLevel, Binary.fromReusedByteArray(bytes));
+    }
+  }
+
+  private static class ArrayDataWriter<E> extends RepeatedWriter<ArrayData, E> {
+    private final DataType elementType;
+
+    private ArrayDataWriter(int definitionLevel, int repetitionLevel,
+                            ParquetValueWriter<E> writer, DataType elementType) {
+      super(definitionLevel, repetitionLevel, writer);
+      this.elementType = elementType;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    protected Iterator<E> elements(ArrayData list) {
+      return new Iterator<E>() {
+        private final int size = list.numElements();
+        private int index = 0;
+
+        @Override
+        public boolean hasNext() {
+          return index != size;
+        }
+
+        @Override
+        public E next() {
+          if (index >= size) {
+            throw new NoSuchElementException();
+          }
+
+          E element;
+          if (list.isNullAt(index)) {
+            element = null;
+          } else {
+            element = (E) list.get(index, elementType);
+          }
+
+          index += 1;
+
+          return element;
+        }
+      };
+    }
+  }
+
+  private static class MapDataWriter<K, V> extends RepeatedKeyValueWriter<MapData, K, V> {
+    private final DataType keyType;
+    private final DataType valueType;
+
+    private MapDataWriter(int definitionLevel, int repetitionLevel,
+                          ParquetValueWriter<K> keyWriter, ParquetValueWriter<V> valueWriter,
+                          DataType keyType, DataType valueType) {
+      super(definitionLevel, repetitionLevel, keyWriter, valueWriter);
+      this.keyType = keyType;
+      this.valueType = valueType;
+    }
+
+    @Override
+    protected Iterator<Map.Entry<K, V>> pairs(MapData map) {
+      return new Iterator<Map.Entry<K, V>>() {
+        private final int size = map.numElements();
```

Comment:
At least to me it's a little harder to read inner classes with fields if they're anonymous like this. Can we separate this `Iterator` out to its own class?